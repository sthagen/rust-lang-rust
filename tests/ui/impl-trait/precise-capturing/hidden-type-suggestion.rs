#![feature(precise_capturing)]

fn lifetime<'a, 'b>(x: &'a ()) -> impl Sized + use<'b> {
//~^ HELP add `'a` to the `use<...>` bound
    x
//~^ ERROR hidden type for
}

fn param<'a, T>(x: &'a ()) -> impl Sized + use<T> {
//~^ HELP add `'a` to the `use<...>` bound
    x
//~^ ERROR hidden type for
}

fn empty<'a>(x: &'a ()) -> impl Sized + use<> {
//~^ HELP add `'a` to the `use<...>` bound
    x
//~^ ERROR hidden type for
}

trait Captures<'a> {}
impl<T> Captures<'_> for T {}

fn missing<'a, 'captured, 'not_captured, Captured>(x: &'a ()) -> impl Captures<'captured> {
//~^ HELP add a `use<...>` bound
    x
//~^ ERROR hidden type for
}

fn main() {}
