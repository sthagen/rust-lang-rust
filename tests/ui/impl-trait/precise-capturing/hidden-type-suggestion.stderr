error[E0700]: hidden type for `impl Sized` captures lifetime that does not appear in bounds
  --> $DIR/hidden-type-suggestion.rs:5:5
   |
LL | fn lifetime<'a, 'b>(x: &'a ()) -> impl Sized + use<'b> {
   |             --                    -------------------- opaque type defined here
   |             |
   |             hidden type `&'a ()` captures the lifetime `'a` as defined here
LL |
LL |     x
   |     ^
   |
help: add `'a` to the `use<...>` bound to explicitly capture it
   |
LL | fn lifetime<'a, 'b>(x: &'a ()) -> impl Sized + use<'b, 'a> {
   |                                                      ++++

error[E0700]: hidden type for `impl Sized` captures lifetime that does not appear in bounds
  --> $DIR/hidden-type-suggestion.rs:11:5
   |
LL | fn param<'a, T>(x: &'a ()) -> impl Sized + use<T> {
   |          --                   ------------------- opaque type defined here
   |          |
   |          hidden type `&'a ()` captures the lifetime `'a` as defined here
LL |
LL |     x
   |     ^
   |
help: add `'a` to the `use<...>` bound to explicitly capture it
   |
LL | fn param<'a, T>(x: &'a ()) -> impl Sized + use<'a, T> {
   |                                                +++

error[E0700]: hidden type for `impl Sized` captures lifetime that does not appear in bounds
  --> $DIR/hidden-type-suggestion.rs:17:5
   |
LL | fn empty<'a>(x: &'a ()) -> impl Sized + use<> {
   |          --                ------------------ opaque type defined here
   |          |
   |          hidden type `&'a ()` captures the lifetime `'a` as defined here
LL |
LL |     x
   |     ^
   |
help: add `'a` to the `use<...>` bound to explicitly capture it
   |
LL | fn empty<'a>(x: &'a ()) -> impl Sized + use<'a> {
   |                                             ++

error[E0700]: hidden type for `impl Captures<'captured>` captures lifetime that does not appear in bounds
  --> $DIR/hidden-type-suggestion.rs:26:5
   |
LL | fn missing<'a, 'captured, 'not_captured, Captured>(x: &'a ()) -> impl Captures<'captured> {
   |            --                                                    ------------------------ opaque type defined here
   |            |
   |            hidden type `&'a ()` captures the lifetime `'a` as defined here
LL |
LL |     x
   |     ^
   |
help: add a `use<...>` bound to explicitly capture `'a`
   |
LL | fn missing<'a, 'captured, 'not_captured, Captured>(x: &'a ()) -> impl Captures<'captured> + use<'captured, 'a, Captured> {
   |                                                                                           ++++++++++++++++++++++++++++++

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0700`.
